libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,177 @@
+	// They can only differ during data recovery / page size migration.
+	pageSize uint32
+
+	// The oldest position in the queue. This is advanced as we receive ACKs from
+	// downstream consumers indicating it is safe to remove old events.
+	firstPosition bufferPosition
+
+	// The position of the next (unwritten) byte in the queue buffer. When an
+	// event is added to the queue, this position is advanced to point to the
+	// first byte after its end.
+	lastPosition bufferPosition
+
+	// The maximum number of pages that can be used for the queue buffer.
+	// This is derived by dividing Settings.MaxBufferSize by pageSize and
+	// rounding down.
+	maxPageCount uint64
+
+	// The number of pages currently occupied by the queue buffer. This can't
+	// be derived from firstPosition and lastPosition because segment length
+	// varies with the size of their last event.
+	// This can be greater than maxPageCount if the maximum buffer size is
+	// reduced on an already-full queue.
+	allocatedPageCount uint64
+}
+
+// diskQueue is the internal type representing a disk-based implementation
+// of queue.Queue.
+type diskQueue struct {
+	settings Settings
+
+	// The persistent queue state. After a filesystem sync this should be
+	// identical to the queue's metadata file.
+	state diskQueueState
+
+	// The position of the next event to read from the queue. If this equals
+	// state.lastPosition, then there are no events left to read.
+	// This is initialized to state.firstPosition, but generally the two differ:
+	// readPosition is advanced when an event is read, but firstPosition is
+	// only advanced when the event has been read _and_ its consumer receives
+	// an acknowledgement (meaning it has been transmitted and can be removed
+	// from the queue).
+	// This is part of diskQueue and not diskQueueState since it represents
+	// in-memory state that should not persist through a restart.
+	readPosition bufferPosition
+}
+
+func init() {
+	queue.RegisterQueueType(
+		"disk",
+		queueFactory,
+		feature.MakeDetails(
+			"Disk queue",
+			"Buffer events on disk before sending to the output.",
+			feature.Beta))
+}
+
+// queueFactory matches the queue.Factory type, and is used to add the disk
+// queue to the registry.
+func queueFactory(
+	ackListener queue.ACKListener, logger *logp.Logger, cfg *common.Config,
+) (queue.Queue, error) {
+	settings, err := SettingsForUserConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	settings.Logger = logger
+	// For now, incoming messages are acked when they are written to disk
+	// (rather than transmitted to the output, as with the memory queue). This
+	// can produce unexpected behavior in some contexts and we might want to
+	// make it configurable later.
+	settings.WriteToDiskACKListener = ackListener
+	return NewQueue(settings)
+}
+
+// NewQueue returns a disk-based queue configured with the given logger
+// and settings.
+func NewQueue(settings Settings) (queue.Queue, error) {
```

Comment:
nit: full name is `diskqueue.NewQueue`. Just naming it `New` would be enough.

I wonder if we want to export diskQueue, so to return a conrete type here.

Reply:
Well, `NewQueue` is the name used in `memqueue`, so I chose it to match, but either name is fine.

I'd strongly prefer not to return a concrete type, since I just refactored out similar internal type dependencies in the memory queue... it's easy to convert an API that returns an abstract interface to one that returns a concrete type, but potentially hard to do the reverse. I only want to expose types and internals as we have explicit (functional) uses for them.